### DISCUSSION TITLE
[Bug] /tmp/go-build2704884287/b439/skywalking_enhance_producer_async_producer.go:22:2: could not import strings (open : no such file or directory)

### DISCUSSION THREAD

**Discussion by Qiangel** [apache/skywalking#12359] [Q&A] [answered]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Go Agent (apache/skywalking-go)

### What happened

I encountered an error while compiling my project with -toolexec.
```
/tmp/go-build2704884287/b439/skywalking_enhance_producer_async_producer.go:22:2: could not import strings (open : no such file or directory)
2024/06/21 07:14:11 exit status 2
```

### What you expected to happen

How can you avoid this mistake

### How to reproduce

I looked up the`less instrument.log` file and found that it was an error from rocketmq.
![企业微信截图_17189550987769](https://github.com/apache/skywalking/assets/39545327/b72fa606-ca6e-4931-bb80-af363abb6567)

ls
```
importcfg
producer.go
skywalking_agent_core_linker.go
skywalking_agent_core_log_logger.go
skywalking_agent_core_metrics_bridge.go      
skywalking_agent_core_metrics_metrics.go     
skywalking_agent_core_operator_common.go     
skywalking_agent_core_operator_intercepter.go
skywalking_agent_core_operator_invocation.go 
skywalking_agent_core_operator_logger.go     
skywalking_agent_core_operator_metrics.go    
skywalking_agent_core_operator_tools.go      
skywalking_agent_core_operator_tracing.go    
skywalking_agent_core_tools_os.go
skywalking_agent_core_tools_reflect.go       
skywalking_agent_core_tools_reflect_bridge.go
skywalking_agent_core_tools_strconv.go       
skywalking_agent_core_tools_sync.go
skywalking_agent_core_tracing_api.go
skywalking_agent_core_tracing_bridge.go      
skywalking_agent_core_tracing_context.go
skywalking_agent_core_tracing_span.go
skywalking_delegator.go
skywalking_enhance_producer_async_producer.go
skywalking_enhance_producer_general_producer.go
skywalking_enhance_producer_send_one_way.go
skywalking_enhance_producer_send_status.go
skywalking_enhance_producer_sync_producer.go
```

less -N skywalking_enhance_producer_general_producer.go
``` go
     19 import (
     20         "strings"
     21         "github.com/apache/rocketmq-client-go/v2/primitive"
     22 )
     23
     24 const (
     25         skywalking_operatorVarProducerrmqSyncSendPrefix         = "RocketMQ/"
     26         skywalking_operatorVarProducerrmqProducerSuffix         = "/Producer"
     27         skywalking_operatorVarProducersemicolon                 = ";"
     28         skywalking_operatorVarProducerrmqSyncComponentID        = 38
     29 )
     30
     31 func skywalking_operatorStaticMethodProducerGeneralProducerBeforeInvoke(invocation sky     31 walking_operatorTypeOperatorInvocation) error {
     32         defaultProducer := invocation.CallerInstance().(*defaultProducer)
     33         peer := strings.Join(defaultProducer.client.GetNameSrv().AddrList(), skywalkin     33 g_operatorVarProducersemicolon)
```

less importcfg
```
# import config
packagefile bytes=/tmp/go-build2001304399/b008/_pkg_.a
packagefile context=/tmp/go-build2001304399/b084/_pkg_.a
packagefile fmt=/tmp/go-build2001304399/b046/_pkg_.a
packagefile github.com/apache/rocketmq-client-go/v2/errors=/tmp/go-build2001304399/b416/_pkg_.a
packagefile github.com/apache/rocketmq-client-go/v2/internal=/tmp/go-build2001304399/b417/_pkg_.a
packagefile github.com/apache/rocketmq-client-go/v2/internal/remote=/tmp/go-build2001304399/b418/_pkg_.a
packagefile github.com/apache/rocketmq-client-go/v2/internal/utils=/tmp/go-build2001304399/b420/_pkg_.a
packagefile github.com/apache/rocketmq-client-go/v2/primitive=/tmp/go-build2001304399/b419/_pkg_.a
packagefile github.com/apache/rocketmq-client-go/v2/rlog=/tmp/go-build2001304399/b421/_pkg_.a 
packagefile github.com/pkg/errors=/tmp/go-build2001304399/b230/_pkg_.a
packagefile github.com/satori/go.uuid=/tmp/go-build2001304399/b440/_pkg_.a
packagefile hash/fnv=/tmp/go-build2001304399/b199/_pkg_.a
packagefile math/rand=/tmp/go-build2001304399/b093/_pkg_.a
packagefile strconv=/tmp/go-build2001304399/b036/_pkg_.a
packagefile sync=/tmp/go-build2001304399/b027/_pkg_.a
packagefile sync/atomic=/tmp/go-build2001304399/b029/_pkg_.a
packagefile time=/tmp/go-build2001304399/b054/_pkg_.a
importcfg (END)
```

I did not do any processing, this plug-in is also own, why the report package does not exist?
What should I do to avoid this mistake

### Are you willing to submit a pull request to fix on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by mrproliu** [comment, marked as answer]
Are you developing a new plugin? For now, the `rocketmq-client-go` framework we haven't enhanced yet. 

For this problem is that the package you enhanced not import `strings` package, so you cannot import it. If you want to know why deeply, please read the [key principle documentation](https://skywalking.apache.org/docs/skywalking-go/next/en/concepts-and-designs/key-principles/). That's why we develop [some tools in the agent API](https://github.com/apache/skywalking-go/tree/main/plugins/core/tools), you cannot invoke the method that package not import. 

In your case, you can ignore the `strings` package and do it my self. Or you can create it in the agent API, and implement it.

**Reply by Qiangel** [reply]
I didn't develop new plugin myself, it's already implemented at 0.4.0
[rockmq](https://github.com/apache/skywalking-go/blob/main/plugins/rocketmq/consumer/consumer.go)

I really don't understand that part of it ,but I would really like to know how to introduce this package. What should I do about it

**Reply by mrproliu** [reply]
> I didn't develop new plugin myself, it's already implemented at 0.4.0
> [rockmq](https://github.com/apache/skywalking-go/blob/main/plugins/rocketmq/consumer/consumer.go)

Oh, Sorry, I might have made a mistake. Also, could you please let me know the current version of RocketMQ being used?

> I really don't understand that part of it ,but I would really like to know how to introduce this package. What should I do about it

What do you mean introduce this package? Do you mean import the `strings` package to the `rocketmq`, you can't because it belongs to the thrid part framework. Also, we have already tested the `v2.1.2` and it works well. In a simple way, you can just downgrade this rocketMQ client to the `v2.1.2`.

**Reply by Qiangel** [reply]
rocketmq version  `github.com/apache/rocketmq-client-go/v2 v2.1.1`.

The problem is, this file is wrong
[plugins/rocketmq/consumer/consumer.go:20](https://github.com/apache/skywalking-go/blob/main/plugins/rocketmq/consumer/consumer.go)

I read the link you sent
 [key principle documentation](https://skywalking.apache.org/docs/skywalking-go/next/en/concepts-and-designs/key-principles/)
![image](https://github.com/apache/skywalking/assets/39545327/bb233dcb-e056-41c4-846f-e1265d82b232)
If I solve it myself, I can just add `strings` to this file? `skywalking-go/imports.go`

**Reply by wu-sheng** [reply]
@mrproliu How is this kind of errors passed our plugin tests?

**Reply by mrproliu** [reply]
We haven't tested the `v2.1.1` version. I just looking at the rocketmq framework code, looks like the they [adding the package import in the v2.1.2](https://github.com/apache/rocketmq-client-go/blob/v2.1.2/producer/option.go), but the [strings package not import in the v2.1.1](https://github.com/apache/rocketmq-client-go/blob/v2.1.1/producer/option.go).

**Reply by mrproliu** [reply]
> If I solve it myself, I can just add strings to this file? skywalking-go/imports.go

No, you can't. Just I sait, it's related to the rocketMQ framework enhancement, not agent framework enhancement. Could you trying to upgrade the version to the `v2.1.2`?

**Reply by Qiangel** [reply]
> We haven't tested the `v2.1.1` version. I just looking at the rocketmq framework code, looks like the they [adding the package import in the v2.1.2](https://github.com/apache/rocketmq-client-go/blob/v2.1.2/producer/option.go), but the [strings package not import in the v2.1.1](https://github.com/apache/rocketmq-client-go/blob/v2.1.1/producer/option.go).

I see..
So in this situation, is it necessary to introduce this strings package in the original way for our plugin to use it? 
Can I solve this problem by modifying the go agent?

**Reply by Qiangel** [reply]
> > If I solve it myself, I can just add strings to this file? skywalking-go/imports.go
> 
> No, you can't. Just I sait, it's related to the rocketMQ framework enhancement, not agent framework enhancement. Could you trying to upgrade the version to the `v2.1.2`?

okk,ty

**Reply by mrproliu** [reply]
Sure, you can modify these two files: https://github.com/apache/skywalking-go/blob/main/plugins/rocketmq/producer/async_producer.go and https://github.com/apache/skywalking-go/blob/main/plugins/rocketmq/producer/general_producer.go to replace the `strings.Join` as simple way(such as do it by self).

**Reply by Qiangel** [reply]
> Sure, you can modify these two files: https://github.com/apache/skywalking-go/blob/main/plugins/rocketmq/producer/async_producer.go and https://github.com/apache/skywalking-go/blob/main/plugins/rocketmq/producer/general_producer.go to replace the `strings.Join` as simple way(such as do it by self).

yes,i know this,ty y answer.

**Reply by mrproliu** [reply]
Do you mean you will create a new PR to support the older version?

**Reply by Qiangel** [reply]
> Do you mean you will create a new PR to support the older version?

hhh, I'd be happy to contribute if I could, but I have little experience and this place is pretty easy.

Speaking of which, I also have a redis cluster, `peer==""` err  I don't know if I should ask you..

**Reply by mrproliu** [reply]
You can create other discuss about the redis plugin.

**Comment by wu-sheng** [comment]
Moving to discussion as this is not an existing feature.
